reactor-netty-examples/src/main/java/reactor/netty/examples/tcp/securechat/SecureChatServer.java

```diff
@@ -0,0 +1,107 @@
+import java.net.UnknownHostException;
+import java.security.cert.CertificateException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.net.ssl.SSLSession;
+
+import io.netty.channel.ChannelId;
+import io.netty.handler.codec.DelimiterBasedFrameDecoder;
+import io.netty.handler.codec.Delimiters;
+import io.netty.handler.ssl.SslHandler;
+import io.netty.handler.ssl.util.SelfSignedCertificate;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.netty.Connection;
+import reactor.netty.ConnectionObserver.State;
+import reactor.netty.tcp.TcpServer;
+import reactor.netty.tcp.TcpSslContextSpec;
+
+public class SecureChatServer {
+
+	private static final int PORT = Integer.parseInt(System.getProperty("port", "8992"));
+
+	private static final boolean WIRETAP = System.getProperty("wiretap") != null;
+
+	public static void main(String[] args) throws UnknownHostException, CertificateException {
+		ConcurrentHashMap<ChannelId, Connection> conns = new ConcurrentHashMap<>();
+		String hostname = InetAddress.getLocalHost().getHostName();
+
+		TcpServer server = TcpServer.create()
+				.port(PORT)
+				.wiretap(WIRETAP)
+				.doOnConnection(connection -> {
+					// cache the new connection. it'll needed later
+					// when the server broadcasts messages from other clients.
+					ChannelId id = connection.channel().id();
+					conns.put(id, connection);
+					connection.addHandlerLast(new DelimiterBasedFrameDecoder(8192, Delimiters.lineDelimiter()));
+				}).childObserve((connection, newState) -> {
+					if (newState == State.DISCONNECTING) {
+						// when a connection is disconnecting,
+						// remove it from the map. We don't want to hold onto
+						// a disconnected connection.
+						ChannelId id = connection.channel().id();
+						conns.remove(id);
+					}
+				}).handle((in, out) -> {
+					List<String> welcomeTexts = new ArrayList<>();
+					welcomeTexts.add("Welcome to " + hostname + " secure chat service!\n");
+					in.withConnection(connection -> {
+						SslHandler handler = connection.channel().pipeline().get(SslHandler.class);
+						SSLSession session = handler.engine().getSession();
+						String cipherSuite = session.getCipherSuite();
+						String msg = "Your session is protected by " + cipherSuite + " cipher suite.\n";
+						welcomeTexts.add(msg);
+					});
+
+					Flux<String> welcomeFlux = Flux.fromIterable(welcomeTexts);
+
+					Flux<String> flux = in.receive()
+							.asString()
+							.takeUntil("bye"::equalsIgnoreCase)
+							.handle((text, sink) -> {
+								in.withConnection(current -> {
+									for (Connection conn : conns.values()) {
+										if (conn == current) {
+											sink.next(text);
+										}
+										else {
+											String msg = "[" + conn.channel().remoteAddress() + "] " + text + '\n';
+
+											conn.outbound().sendString(Mono.just(msg)).then().subscribe();
+										}
+									}
+								});
```

Comment:
```suggestion
                                in.withConnection(current -> Flux.fromStream(conns.values()
                                                .stream()
                                                .filter(conn -> conn != current))
                                        .map(conn -> conn.outbound()
                                                .sendString(Mono.just("[" + conn.channel().remoteAddress() + "] " + text + '\n')).then())
                                        .doOnComplete(() -> sink.next(text))
                                        .subscribe()
                                );
```

in order to encourage back pressure, what about waiting for the flux of forwarded messages to be completed before returning the message to the original user, something that would look like above ?

Reply:
I just changed the above proposed change, please take care (I replaced map by flatMap line 86).
thanks.

Reply:
mmm, my suggestion does not seem to work, please ignore this one for the moment. sorry.

Reply:
`handle` is using a `SynchronousSInk`, that's why the above was not working.